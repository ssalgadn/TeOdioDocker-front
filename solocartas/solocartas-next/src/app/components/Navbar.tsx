'use client';

import Link from 'next/link';
import { UserCircle2, ChevronDown, Sun, Moon } from 'lucide-react';
import { useTheme } from 'next-themes';
import { usePathname } from 'next/navigation';
import { useState, useRef, useEffect } from 'react';

// ... (La interfaz DropdownItem y el array megaMenuColumns no cambian)
interface DropdownItem {
  name: string;
  href?: string;
  subItems?: DropdownItem[];
}

const megaMenuColumns: DropdownItem[] = [
  {
    name: 'Pokémon',
    href: '/products?game=pokemon',
    subItems: [
      { name: 'Boosters', href: '/products?game=pokemon&type=boosters' },
      { name: 'Bundles', href: '/products?game=pokemon&type=bundles' },
      { name: 'Singles', href: '/products?game=pokemon&type=singles' },
    ],
  },
  {
    name: 'Yu-Gi-Oh!',
    href: '/products?game=yu-gi-oh',
    subItems: [
      { name: 'Boosters', href: '/products?game=yu-gi-oh&type=boosters' },
      { name: 'Bundles', href: '/products?game=yu-gi-oh&type=bundles' },
      { name: 'Singles', href: '/products?game=yu-gi-oh&type=singles' },
    ],
  },
  {
    name: 'Magic: The Gathering',
    href: '/products?game=magic-the-gathering',
    subItems: [
      { name: 'Boosters', href: '/products?game=magic-the-gathering&type=boosters' },
      { name: 'Bundles', href: '/products?game=magic-the-gathering&type=bundles' },
      { name: 'Singles', href: '/products?game=magic-the-gathering&type=singles' },
    ],
  },
];


export default function Navbar() {
  const pathname = usePathname();
  const [isUserDropdownOpen, setIsUserDropdownOpen] = useState(false);
  const [isCatalogOpen, setCatalogOpen] = useState(false);
  const { theme, setTheme } = useTheme();
  const [mounted, setMounted] = useState(false);

  const userDropdownRef = useRef<HTMLDivElement>(null);
  const catalogCloseTimer = useRef<NodeJS.Timeout | null>(null);

  useEffect(() => setMounted(true), []);

  useEffect(() => {
    function handleClickOutside(event: MouseEvent) {
      if (userDropdownRef.current && !userDropdownRef.current.contains(event.target as Node)) {
        setIsUserDropdownOpen(false);
      }
    }
    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, []);

  const handleLogout = () => {
    console.log('Logout clicked');
    setIsUserDropdownOpen(false);
  };
  
  const handleCatalogEnter = () => {
    if (catalogCloseTimer.current) {
      clearTimeout(catalogCloseTimer.current);
    }
    setCatalogOpen(true);
  };

  const handleCatalogLeave = () => {
    catalogCloseTimer.current = setTimeout(() => {
      setCatalogOpen(false);
    }, 200); // Pequeño delay para permitir que el mouse se mueva al menú
  };

  return (
    // La NAV ahora es el contenedor principal para la lógica de hover
    <div className="relative" onMouseLeave={handleCatalogLeave}>
      <nav className="bg-white dark:bg-gray-900 shadow sticky top-0 z-50">
        <div className="container mx-auto px-6 py-3 flex justify-between items-center">
          {/* Left Side: Logo and Navigation Links */}
          <div className="flex items-center gap-8">
            <Link href="/">
              <span className="text-2xl font-bold text-blue-700 dark:text-blue-500 tracking-tight">SoloCartas</span>
            </Link>
            <div className="hidden md:flex gap-6 items-center">
              <Link
                href="/"
                className={`font-medium ${pathname === '/' ? 'text-blue-600 dark:text-blue-400' : 'text-gray-700 dark:text-gray-300'} hover:text-blue-600 dark:hover:text-blue-400 transition-colors`}
              >
                Inicio
              </Link>
              
              {/* --- CAMBIO AQUÍ: Se reemplazó el <div> por <Link> --- */}
              <Link
                href="/products"
                onMouseEnter={handleCatalogEnter}
                className={`font-medium ${pathname.startsWith('/products') || isCatalogOpen ? 'text-blue-600 dark:text-blue-400' : 'text-gray-700 dark:text-gray-300'} hover:text-blue-600 dark:hover:text-blue-400 transition-colors flex items-center cursor-pointer`}
              >
                <span>Catálogo</span>
                <ChevronDown className={`h-4 w-4 ml-1 text-current transition-transform ${isCatalogOpen ? 'rotate-180' : ''}`} />
              </Link>
            </div>
          </div>

          {/* Right Side: Theme Toggle and User Dropdown */}
          <div className="flex items-center gap-4">
            {mounted && (
              <button
                onClick={() => setTheme(theme === 'dark' ? 'light' : 'dark')}
                className="p-2 rounded-full text-gray-600 dark:text-gray-400 hover:bg-gray-200 dark:hover:bg-gray-700"
                aria-label="Toggle theme"
              >
                {theme === 'dark' ? <Sun size={20} /> : <Moon size={20} />}
              </button>
            )}
            <div className="relative" ref={userDropdownRef}>
              <button onClick={() => setIsUserDropdownOpen(!isUserDropdownOpen)} className="flex items-center">
                <UserCircle2 size={36} className="text-gray-500 dark:text-gray-400" strokeWidth={1.5} />
                <ChevronDown className={`h-5 w-5 ml-1 text-gray-600 dark:text-gray-400 transition-transform ${isUserDropdownOpen ? 'rotate-180' : ''}`} />
              </button>
              {isUserDropdownOpen && (
                <div className="absolute right-0 mt-2 w-48 bg-white dark:bg-gray-800 rounded-md shadow-xl py-1 z-50">
                  <Link href="/login" onClick={() => setIsUserDropdownOpen(false)} className="block px-4 py-2 text-sm text-gray-700 dark:text-gray-200 hover:bg-gray-100 dark:hover:bg-gray-700">Login</Link>
                  <Link href="/register" onClick={() => setIsUserDropdownOpen(false)} className="block px-4 py-2 text-sm text-gray-700 dark:text-gray-200 hover:bg-gray-100 dark:hover:bg-gray-700">Register</Link>
                  <button onClick={handleLogout} className="block w-full text-left px-4 py-2 text-sm text-gray-700 dark:text-gray-200 hover:bg-gray-100 dark:hover:bg-gray-700">Logout</button>
                </div>
              )}
            </div>
          </div>
        </div>
      </nav>

      {/* Mega Menu está AHORA FUERA de la NAV pero dentro del wrapper con la lógica de hover */}
      <div
        onMouseEnter={handleCatalogEnter} // Mantener abierto si el mouse entra al menú
        className={`absolute top-full left-0 right-0 // Posicionamiento full-width
                   bg-gray-700 dark:bg-gray-800 shadow-lg z-40
                   transition-opacity duration-300
                   ${isCatalogOpen ? 'opacity-100' : 'opacity-0 pointer-events-none'}`}
      >
        <div className="container mx-auto px-6 py-8">
          <div className="grid grid-cols-1 md:grid-cols-3 gap-x-8 gap-y-10">
            {megaMenuColumns.map((column) => (
              <div key={column.name} className="space-y-4">
                <Link
                  href={column.href || '#'}
                  onClick={() => setCatalogOpen(false)}
                  className="text-lg font-semibold text-gray-100 dark:text-gray-50 hover:text-blue-300 dark:hover:text-blue-300 transition-colors block mb-3"
                >
                  {column.name}
                </Link>
                {column.subItems && (
                  <ul className="space-y-2">
                    {column.subItems.map((subItem) => (
                      <li key={subItem.name}>
                        <Link
                          href={subItem.href || '#'}
                          onClick={() => setCatalogOpen(false)}
                          className="text-gray-300 dark:text-gray-400 hover:text-gray-100 dark:hover:text-gray-50 transition-colors text-sm"
                        >
                          {subItem.name}
                        </Link>
                      </li>
                    ))}
                  </ul>
                )}
              </div>
            ))}
          </div>
        </div>
      </div>
    </div>
  );
}