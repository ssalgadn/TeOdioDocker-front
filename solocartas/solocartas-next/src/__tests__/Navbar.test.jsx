import { render } from '@testing-library/react';
import Navbar from '@/app/components/Navbar';

jest.mock('next-themes', () => ({ useTheme: () => ({ theme: 'light', setTheme: jest.fn() }) }));
jest.mock('next/navigation', () => ({ usePathname: () => '/', useRouter: () => ({ push: jest.fn() }) }));

jest.mock('@/contexts/AuthContext', () => ({
  useAuth: () => ({ isAuthenticated: false, user: null, isLoading: false, logout: jest.fn() }),
}));

jest.mock('next/link', () => {
  return ({ children, href }) => <a href={href}>{children}</a>;
});

jest.mock('next/image', () => (props) => {
  // eslint-disable-next-line @next/next/no-img-element
  return <img {...props} alt={props.alt} />;
});

describe('Navbar snapshot', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Navbar />);
    expect(asFragment()).toMatchSnapshot();
  });
});
