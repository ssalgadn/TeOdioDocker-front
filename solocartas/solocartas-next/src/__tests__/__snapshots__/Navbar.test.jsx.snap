// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar snapshot matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="sticky top-0 z-50 bg-white dark:bg-slate-900 shadow-sm border-b border-gray-200 dark:border-slate-700"
  >
    <nav
      class="container mx-auto px-6 py-3 flex justify-between items-center"
    >
      <div
        class="flex items-center gap-8"
      >
        <a
          href="/"
        >
          SoloCartas
        </a>
        <div
          class="hidden md:flex items-center gap-6 text-gray-700 dark:text-gray-300 font-medium"
        >
          <a
            href="/"
          >
            Inicio
          </a>
          <a
            href="/about-us"
          >
            About Us
          </a>
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="flex items-center gap-1 cursor-pointer hover:text-blue-600 dark:hover:text-blue-400 transition-colors font-medium focus:outline-none "
            type="button"
          >
            Catálogo
            <svg
              aria-hidden="true"
              class="lucide lucide-chevron-down h-4 w-4 transition-transform "
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m6 9 6 6 6-6"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="flex items-center gap-4"
      >
        <button
          aria-label="Toggle theme"
          class="p-2 rounded-full text-gray-600 dark:text-gray-400 hover:bg-gray-200 dark:hover:bg-gray-700 transition"
          type="button"
        >
          <svg
            class="lucide lucide-moon"
            fill="none"
            height="20"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
            />
          </svg>
        </button>
        <div
          class="relative"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="flex items-center gap-1 focus:outline-none"
            type="button"
          >
            <svg
              class="lucide lucide-circle-user-round text-gray-500 dark:text-gray-400"
              fill="none"
              height="36"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              width="36"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 20a6 6 0 0 0-12 0"
              />
              <circle
                cx="12"
                cy="10"
                r="4"
              />
              <circle
                cx="12"
                cy="12"
                r="10"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="lucide lucide-chevron-down h-5 w-5 text-gray-600 dark:text-gray-400 transition-transform "
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m6 9 6 6 6-6"
              />
            </svg>
          </button>
        </div>
      </div>
    </nav>
    <div
      class="absolute top-full left-0 right-0 bg-gray-700 dark:bg-gray-800 shadow-lg z-40 transition-opacity duration-300 opacity-0 pointer-events-none"
    >
      <div
        class="container mx-auto px-6 py-8"
      >
        <div
          class="grid grid-cols-1 md:grid-cols-3 gap-x-8 gap-y-10"
        >
          <div
            class="space-y-4"
          >
            <a
              href="/products?game=pokemon"
            >
              Pokémon
            </a>
            <ul
              class="space-y-2"
            >
              <li>
                <a
                  href="/products?game=pokemon&product_type=booster"
                >
                  Boosters
                </a>
              </li>
              <li>
                <a
                  href="/products?game=pokemon&product_type=bundle"
                >
                  Bundles
                </a>
              </li>
              <li>
                <a
                  href="/products?game=pokemon&product_type=single"
                >
                  Singles
                </a>
              </li>
            </ul>
          </div>
          <div
            class="space-y-4"
          >
            <a
              href="/products?game=yu-gi-oh"
            >
              Yu-Gi-Oh!
            </a>
            <ul
              class="space-y-2"
            >
              <li>
                <a
                  href="/products?game=yu-gi-oh&product_type=booster"
                >
                  Boosters
                </a>
              </li>
              <li>
                <a
                  href="/products?game=yu-gi-oh&product_type=bundle"
                >
                  Bundles
                </a>
              </li>
              <li>
                <a
                  href="/products?game=yu-gi-oh&product_type=single"
                >
                  Singles
                </a>
              </li>
            </ul>
          </div>
          <div
            class="space-y-4"
          >
            <a
              href="/products?game=magic-the-gathering"
            >
              Magic: The Gathering
            </a>
            <ul
              class="space-y-2"
            >
              <li>
                <a
                  href="/products?game=magic-the-gathering&product_type=booster"
                >
                  Boosters
                </a>
              </li>
              <li>
                <a
                  href="/products?game=magic-the-gathering&product_type=bundle"
                >
                  Bundles
                </a>
              </li>
              <li>
                <a
                  href="/products?game=magic-the-gathering&product_type=single"
                >
                  Singles
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
